spec/ffg.tla

```diff
@@ -87,4 +95,202 @@ get_slashable_nodes_unoptimized(vote_view) ==
+\* if at least two-thirds of the total validator set's weight supports the transition from this `checkpoint` to the next. Specifically, it first checks if the `checkpoint` is justified using
+\* `is_justified_checkpoint(checkpoint, node_state)`. Then it retrieves the set of validators and their balances for the slot associated with the `checkpoint`.
+\* This is done through `get_validator_set_for_slot`. Then it calculates the total weight (stake) of validators who have cast votes
+\* linking the `checkpoint` to the next slot, using `get_validators_in_FFG_votes_linking_to_a_checkpoint_in_next_slot` to identify these validators
+\* and `validator_set_weight` to sum their stakes. Finally it checks if `FFG_support_weight * 3 >= tot_validator_set_weight * 2` to finalize `checkpoint`.
+\* @type: ($checkpoint, $commonNodeState) => Bool;
+is_finalized_checkpoint(checkpoint, node_state) ==
+    IF ~is_justified_checkpoint(checkpoint, node_state)
+    THEN FALSE
+    ELSE 
+        LET 
+            validatorBalances == 
+                GET_VALIDATOR_SET_FOR_SLOT(
+                    get_block_from_hash(checkpoint.block_hash, node_state), 
+                    checkpoint.block_slot, 
+                    node_state
+                )
+        IN LET
+            FFG_support_weight == 
+                validator_set_weight(
+                    get_validators_in_FFG_votes_linking_to_a_checkpoint_in_next_slot(checkpoint, node_state), 
+                    validatorBalances
+                )
+            tot_validator_set_weight == 
+                validator_set_weight(DOMAIN validatorBalances, validatorBalances)
+        IN FFG_support_weight * 3 >= tot_validator_set_weight * 2
+        
+    
+\* SRC: https://github.com/saltiniroberto/ssf/blob/ad3ba2c21bc1cd554a870a6e0e4d87040558e129/high_level/common/ffg.py#L181
+\* @type: (Set($checkpoint), $commonNodeState) => Set($checkpoint);
+filter_out_non_finalized_checkpoint(checkpoints, node_state) ==
+    {checkpoint \in checkpoints: is_finalized_checkpoint(checkpoint, node_state)}
+
+\* SRC: https://github.com/saltiniroberto/ssf/blob/7ea6e18093d9da3154b4e396dd435549f687e6b9/high_level/common/ffg.py#L10
+\* It extracts a set of ffg targets from a set of `votes`.
+\* @type: (Set($signedVoteMessage)) => Set($checkpoint);
+get_set_FFG_targets(votes) ==
+    { vote.message.ffg_target: vote \in votes }
+
+\* SRC: https://github.com/saltiniroberto/ssf/blob/7ea6e18093d9da3154b4e396dd435549f687e6b9/high_level/common/ffg.py#L188
+\* It retrieves from `node_state` all the checkpoints that have been finalized.
+\* @type: ($commonNodeState) => Set($checkpoint); 
+get_finalized_checkpoints(node_state) ==
+    LET ffgTargets == get_set_FFG_targets(node_state.view_votes)
+    IN LET filtered == filter_out_non_finalized_checkpoint(ffgTargets, node_state)
+    IN filtered \cup {genesis_checkpoint(node_state)}
```

Comment:
@konnov requested somewhere that we write `\union` and `\intersect`, and I agree, it's easier to read (and less prone to typos).

```suggestion
    IN filtered \union {genesis_checkpoint(node_state)}
```